### PR TITLE
localStorage for code + capturing console errors

### PR DESCRIPTION
**Pull request by austinnguyen2001**
The command to save code is ctrl-s

### CODE REVIEW

File: web-ui/components/EditorPanelComponents/Console.vue [hunk 1]
```diff
@@ -1,11 +1,40 @@
...
+      },
+    },
+  },
+  computed: {
+    computedConsole() {
+      let computedConsole = "";
+      this.console.forEach(line => {
+        computedConsole += `<div class = "console-line ${line.type}">${line.output}</div>`;
+      });
+      return computedConsole;
```
Comment (mdaffin) [inline web-ui/components/EditorPanelComponents/Console.vue]: This all can be replaced with 
```js
return this.console
  .map(line => `<div class="console-line ${line.type}">${line.output}</div>`)
  .join();
```

File: web-ui/components/EditorPanelComponents/Console.vue [hunk 2]
```diff
@@ -1,11 +1,40 @@
...
   },
 };
 </script>
+
+<style>
+.console-line {
+  padding: 5px;
+}
+.console-line:nth-child(2n) {
+  background-color: #e6dfd7;
```
Comment (mdaffin) [inline web-ui/components/EditorPanelComponents/Console.vue]: Not sure about the alternative line coloring, maybe we should tone the contrast down on it.

File: web-ui/pages/index.vue [hunk 3]
```diff
@@ -19,18 +19,33 @@ export default {
...
-    window.log = output => {
-      this.addToLog(output);
-    };
-  },
   data() {
     return {
       code:
-        "const a = 10;\nconst b = 20;\nconsole.log(a + b);\nconsole.log('A String');\n",
-      console: "",
+        "const a = 10;\nconst b = 20;\nconsole.log(a + b);\nconsole.log('A String');\nsdfsdf\n",
```
Comment (mdaffin) [inline web-ui/pages/index.vue]: Remove this invalid code - it is useful for testing but we don't really want the default script to have an error in it.

File: web-ui/pages/index.vue [hunk 4]
```diff
@@ -19,18 +19,33 @@ export default {
...
+      console: [],
     };
   },
+  mounted() {
+    if (typeof Storage !== "undefined")
+      if (localStorage.code) this.code = localStorage.code;
```
Comment (mdaffin) [inline web-ui/pages/index.vue]: I dislike implicit bracers as it is easy to introduce a none obvious an error when refactoring and it keeps things more consistent if the bracers are always used. You can also remove the nested if:

```js
    if (typeof Storage !== "undefined" && localStorage.code) {
      this.code = localStorage.code;
    }
```

File: web-ui/pages/index.vue [hunk 5]
```diff
@@ -19,18 +19,33 @@ export default {
...
+
+    window.log = (output, type) => {
+      this.addToLog(output, type);
+    };
+
+    let storeCode = e => {
```
Comment (mdaffin) [inline web-ui/pages/index.vue]: Prefer `const` over `let` unless you explicitly need to modify the variable - which here you don't.

Also, `saveCode` would be a better name, though you can just inline this function below.

File: web-ui/pages/index.vue [hunk 6]
```diff
@@ -44,11 +59,15 @@ export default {
       const doc = iframe.contentDocument;
 
       const logger = `
-        var console = {
+        let console = {
```
Comment (mdaffin) [inline web-ui/pages/index.vue]: This can be a `const`

File: web-ui/pages/index.vue [hunk 7]
```diff
@@ -44,11 +59,15 @@ export default {
...
           log:(output) => {
-            parent.log(output);
+            parent.log(output, 'norm');
           }
         };
+
+        window.onerror = function(error, url, line) {
+          parent.log('Javascript Error : ' + error+ ' on line ' + (line - 10), 'error');
```
Comment (mdaffin) [inline web-ui/pages/index.vue]: Prefer template literals over string concatination:

```js
`Javascript Error : ${error} on line ${line - 10} error`
```
Also, why `line - 10`? where does this magic number come from? What happens if that changes? Is there a way co calculate the value instead?

**Comment by mdaffin** [comment]
implements #40 (missing tests)
implements #23 (missing tests)
implements most of #37 (missing a save button and tests)

File: web-ui/pages/index.vue [hunk 8]
```diff
@@ -19,18 +19,33 @@ export default {
...
+
+    window.log = (output, type) => {
+      this.addToLog(output, type);
+    };
+
+    let storeCode = e => {
+      if (
+        e.keyCode == 83 &&
+        (navigator.platform.match("Mac") ? e.metaKey : e.ctrlKey)
+      ) {
```
Comment (mdaffin) [inline web-ui/pages/index.vue]: Don't quite like how this line has broken up, might be better to do:
```js
const modifierKey = (navigator.platform.match("Mac") ? e.metaKey : e.ctrlKey);
 if (e.keyCode == 83 && modifierKey) {
```

**Comment by austinnguyen2001** [comment]
the - 10 comes from the height of the logger code.

File: web-ui/pages/index.vue (hunk 7 above)
Comment (mdaffin) [reply web-ui/pages/index.vue]: > the - 10 comes from the height of the logger code.
And if we ever need to change the text it you have to remember to update this magic number. It is very easy to forget or for someone else to not know it even matters.

It is best to use calculated values where you can - though in this case it is a bit awkward given it is self self-referential. One way would be to use a placeholder and `string.replace()`:

```js
const logger = `
...
parent.log('.... LINE_OFFSET ...')
...
`;
const line_count = logger.split(/\r\n|\r|\n/).length;
...
unescape("%3Cscript%3E" + logger.replace("LINE_OFFSET", line_count) + this.code + "%3C/script%3E"),
```

Comment (mdaffin) [reply web-ui/pages/index.vue]: Hmm, I wonder if we use two different script tags if the line count would then need to be off set at all? Might be worth a try:
```js
doc.open();
doc.write(unescape(
  `%3Cscript%3E
  ${logger}
  ${console.log(new Date(Date.now()).toLocaleTimeString())}
  %3C/script%3E`
));
doc.write(unescape(
  `%3Cscript%3E
  ${this.code}
  %3C/script%3E`
));
doc.write(unescape(
  `%3Cscript%3E
  ${console.log('\\n')}
  %3C/script%3E`
));
doc.close();
```

**Pull request merged**